src/App.js

```diff
@@ -0,0 +1,87 @@
+import config from './config';
+import Api from './Api';
+import InfoBarWidget from './InfoBarWidget';
+import { activationInstance } from './ActivationSingleton';
+
+/**
+ * Application class, responsible for running, activating,
+ * and toggling the entire application.
+ *
+ * @class
+ */
+class App {
+	/**
+	 * Only instantiate once, so the initialization doesn't run again
+	 * even if this is called on multiple clicks/calls
+	 *
+	 * @constructor
+	 * @param {App} A class instance
+	 */
+	constructor() {
+		// Instantiate only once
+		if ( !App.instance ) {
+			this.initialized = false;
+			App.instance = this;
+		}
```

Comment:
This system (here and in ActivationSingleton.js) feels a little weird. Instead of preventing reinstantiation within this class, we could prevent it wherever the class is being used, no?

I'm sure I'm missing something, and frankly it seems kind of neat to not have to worry about re-using an existing instance. It just deviates from what I would normally expect a constructor to do (`new Whatever()` should always give me a new `Whatever`)

Reply:
Yeah, this is why it's a singleton; a Singleton only allows for one instantiation, no matter what. 
https://en.wikipedia.org/wiki/Singleton_pattern

So the idea here is that you can import the file and know that you're always using the same instance.

Thta said, I think I can make it by having a shared instance, rather than forcing a singleton. It's more "proper" as a singleton, but I don't know if it's completely needed here. I'll take another look to see if the other method is nicer while keeping up with the needed spec.

Reply:
BTW, this is a super awesome writeup about Javascript design patterns: https://addyosmani.com/resources/essentialjsdesignpatterns/book/

it was written for ES5, but the conceptual design patterns still apply. The code that I used is just the way to do a singleton for an ES6 class.

Reply:
In other languages, a singleton would have a private constructor, and some factory method. I guess JS can't do that, so ensuring the constructor only ever returns the same instance is as good as we get.

Reply:
Yeah, that's how es5-style JS singleton was done (see the activation.singleton.js file I'm overriding with this patch) -- the methods are all private, and you have an exposed method that always gives you the same instance, then calls whatever method is needed.

One of the issues with a singleton is that it's hard to do tests on it, because it's not "pure"; it's always polluted by the previous tests. I ran into this in here too which is why the tests are in this order (if you move the first test to be the second, it'll fail :\ ) I almost made the ActivationSingleton into just a function that shares an instance, but in the actual operation of the system it is supposed to be a Singleton, so I think a small payment of ordering the tests might be worth it.